docs/proposals/19-migrating-to-prometheus-operator.md.md

```diff
@@ -0,0 +1,444 @@
+                regex: cloud-controller-manager;metrics
+              # common metrics
+              - action: labelmap
+                  regex: __meta_kubernetes_service_label_(.+)
+              - source_labels: [ __meta_kubernetes_pod_name ]
+                  target_label: pod
+              metric_relabel_configs:
+              - source_labels: [ __name__ ]
+                regex: ^(rest_client_requests_total|process_max_fds|process_open_fds)$
+                action: keep
+        ```
+
+    - This configmap will be split up into 2 separate resources. One for the
+      `alerting_rules` and another for the `scrape_config`. The `alerting_rules`
+      should be converted into a `PrometheusRules` object. Since the
+      `scrape_config` only has one `job_name` we will only need one
+      `ServiceMonitor` or `PodMonitor` for this. The following is an example of
+      how this could be done. There are multiple ways to get the same results
+      and this is just one example.
+
+        ```yaml
+        apiVersion: monitoring.coreos.com/v1
+        kind: ServiceMonitor
+        metadata:
+          labels:
+            cluster: shoot--project--name
+          name: cloud-controller-manager
+          namespace: shoot--project--name
+        spec:
+          endpoints:
+          - port: metrics # scrape the service port with name `metrics`
+            bearerTokenFile: /var/run/secrets/gardener.cloud/shoot/token/token # could also be replaced with a secret
+            metricRelabelings:
+            - sourceLabels: [ __name__ ]
+              regex: ^(rest_client_requests_total|process_max_fds|process_open_fds)$
+              action: keep
+          namespaceSelector:
+            matchNames:
+            - shoot--project--name
+          selector:
+            matchLabels:
+              role: cloud-controller-manager # discover any service with this label
+        ```
+
+      ```yaml
+      apiVersion: monitoring.coreos.com/v1
+      kind: PrometheusRule
+      metadata:
+        labels:
+          cluster: shoot--project--name
+        name: cloud-controller-manager-rules
+        namespace: shoot--project--name
+      spec:
+        groups:
+        - name: cloud-controller-manager.rules
+          rules:
+          - alert: CloudControllerManagerDown
+            expr: absent(up{job="cloud-controller-manager"} == 1)
+            for: 15m
+            labels:
+              service: cloud-controller-manager
+              severity: critical
+              type: seed
+              visibility: all
+            annotations:
+              description: All infrastruture specific operations cannot be completed (e.g. creating loadbalancers or persistent volumes).
+              summary: Cloud controller manager is down.
+      ```
+
+    - Adding meta monitoring for the extensions: If the extensions need to be
+      scraped for monitoring, the extensions must bring their own [Custom
+      Resources][prom-crds].
```

Comment:
Ideally, in the namespace where the extension is. The challenge here is getting `Prometheus` to discover these resources since they are in other namespaces with random names. If the extension namespaces have labels, then it should be possible with `Prometheus.spec.podMonitorNamespaceSelector` to discover `PodMonitors` for in the extension namespaces for example.

Reply:
Ok, what do you think about the following proposal?
We don't only select `*Monitors` in specific namespaces, but in all namespaces. `*Monitors` are then labeled with some well-known labels that specify which prometheus instance will pick it up. E.g. 
- any namespace, label `prometheus=seed` -> picked up by seed prometheus instance
- shoot namespace, label `prometheus=shoot,target=seed` -> picked up by the respective shoot prometheus instance
- shoot namespace, label `prometheus=shoot,target=shoot` -> picked up by the to-be-introduced shoot prometheus agent instance that can discover targets in the shoot itself

(Label keys/values are just examples)

I think, something like this would drastically simplify discovery of the monitor objects.

Reply:
I think we should do both and I believe we should because otherwise we could add unnecessary network load to our etcd (each Prometheus would list all `*Monitors` and the API server will discard the ones that don't match the labels). For `serviceMonitors` we would have to add a `prometheus.spec.serviceMonitorNamespaceSelector`. However, since we cannot OR on label keys using `matchExpressions` I don't think we could currently monitor the garden namespace and the extensions. If the `garden` namespace had the label `gardener.cloud/role: garden` similar to the extensions' `gardener.cloud/role: extension` then the following would work:

```yaml
serviceMonitorNamespaceSelector:
  matchExpressions:
  - key: 'gardener.cloud/role'
    operator: In
    values: [garden,extensions]
```

Reply:
After a while I realized it is possible to select multiple namespaces since they have a label `kubernetes.io/metadata.name`. Something like this should work.

```yaml
  matchExpressions:
  - key: kubernetes.io/metadata.name
    operator: In
    values:
    - monitoring
    - shoot--project--name
```